include/range/v3/utility/swap.hpp

```diff
@@ -42,10 +53,26 @@ namespace ranges
         /// \cond
         namespace adl_swap_detail
         {
-            using std::swap;
+
+            // The following implementation of swap is adapted from libc++:
+            template <class T>
+            RANGES_CXX14_CONSTEXPR
+            typename std::enable_if<
+                std::is_move_constructible<T>::value &&
+                std::is_move_assignable<T>::value
+             >::type
+            swap(T& x, T& y)
+            noexcept(std::is_nothrow_move_constructible<T>::value &&
+                     std::is_nothrow_move_assignable<T>::value)
+            {
+                T t(std::move(x));
+                x = std::move(y);
+                y = std::move(t);
+            }
```

Comment:
I think you're mistaken about the ambiguity error. I just tried the following with your branch:

~~~~cpp
#include <utility>
#include <complex>
#include <range/v3/core.hpp>

template<typename T>
struct S {
    T t;
};

int main()
{
    S<std::complex<float>> s,t;
    ranges::swap(s,t);
}
~~~~

With clang, this is the error I get:

~~~
/cygdrive/c/Users/eric/Code/range-v3/libs/range/v3/scratch/range.cpp:13:5: error: no matching function for call to object of type 'const ranges::v3::adl_swap_detail::swap_fn'
    ranges::swap(s,t);
    ^~~~~~~~~~~~
/cygdrive/c/Users/eric/Code/range-v3/include/range/v3/utility/swap.hpp:94:17: note: candidate template ignored: substitution failure [with T = S<std::complex<float> > &, U = S<std::complex<float> > &]: no type named 'type' in
      'meta::v1::detail::_if_<std::integral_constant<bool, false> >'
                operator()(T && t, U && u) const noexcept(is_nothrow_swappable<T, U>::value)
                ^
1 error generated.
~~~

With my branch, it compiles fine. Reread my comment above about types with `std` as an associated namespace and ADL. This is the major problem with your approach. I'm quite certain that any effort to redefine `swap` will fail this way.